libyul/optimiser/NameSimplifier.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/NameSimplifier.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+#include <libyul/optimiser/OptimizerUtilities.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <regex>
+
+using namespace solidity::yul;
+using namespace std;
+
+NameSimplifier::NameSimplifier(
+	OptimiserStepContext& _context,
+	Block const& _ast
+):
+	m_context(_context),
+	m_usedNames(_context.reservedIdentifiers)
+{
+	for (YulString name: m_usedNames)
+		m_translations[name] = name;
+
+	set<YulString> allNames = NameCollector(_ast).names();
+	m_usedNames += allNames;
+	for (YulString name: allNames)
+		findSimplification(name);
+}
+
+void NameSimplifier::operator()(FunctionDefinition& _funDef)
+{
+	translate(_funDef.name);
+	renameVariables(_funDef.parameters);
+	renameVariables(_funDef.returnVariables);
+	ASTModifier::operator()(_funDef);
+}
+
+void NameSimplifier::operator()(VariableDeclaration& _varDecl)
+{
+	renameVariables(_varDecl.variables);
+	ASTModifier::operator()(_varDecl);
+}
+
+void NameSimplifier::renameVariables(vector<TypedName>& _variables)
+{
+	for (TypedName& typedName: _variables)
+		translate(typedName.name);
+}
+
+void NameSimplifier::operator()(Identifier& _identifier)
+{
+	translate(_identifier.name);
+}
+
+void NameSimplifier::operator()(FunctionCall& _funCall)
+{
+	// The visitor on its own does not visit the function name.
+	if (!m_context.dialect.builtin(_funCall.functionName.name))
+		(*this)(_funCall.functionName);
+	ASTModifier::operator()(_funCall);
+}
+
+void NameSimplifier::findSimplification(YulString _name)
+{
+	if (m_translations.count(_name))
+		return;
+
+	string name = _name.str();
+
+	static auto replacements = vector<pair<regex, string>>{
+		{regex("_[$]\\d+"), ""}, // removes AST IDs
+		{regex("(abi_..code.*)_to_.*"), "$1"}, // removes _to... for abi functions
+		{regex("(stringliteral_[0-9a-f][0-9a-f][0-9a-f][0-9a-f])[0-9a-f]*"), "$1"}, // shorten string literal
+		{regex("tuple_t_"), ""},
+		{regex("_memory_ptr"), ""},
+		{regex("_calldata_ptr"), "_calldata"},
+		{regex("_fromStack"), ""},
+		{regex("t_contract$_([^_]*)_"), "$1"},
```

Comment:
```suggestion
		{regex("t_contract\\$_([^_]*)_"), "$1"},
```
I think you have to escape the `$` here.  Or put it inside `[$]` like the first regex.